CMS-537 product banner

walsh9 opened the pull request:
Update product banner to use system messages API

This uses [ember-data-model-fragments ](https://github.com/lytics/ember-data-model-fragments)to keep the product banner messages as a separate model object that lives under the system messages model instead of being a json object.

I want to move the banner to its own component, but plan to do that refactor in a later ticket either when implementing the bottom banner, or when moving this to live across pages in the application template. Hopefully the acceptance tests will hold out in either case.

Had some trouble getting mirage to properly ccreate and serialize and return the json i expected using factories so I just put the json i wanted into a response... (i feel like it might not like that `system_messages` api is plural but returns a single object? not sure)

noslouch commented on app/serializers/system-messages.js:
```diff
@@ -0,0 +1,28 @@
+  keyForAttribute(attr) {
+    return underscore(attr);
+  },
+  normalizeFindRecordResponse(store, galleryClass, payload, id) {
```
gallery class?

walsh9 replied:
caught copy and pasting

noslouch commented on app/serializers/system-messages.js:
```diff
@@ -0,0 +1,28 @@
+  normalizeFindRecordResponse(store, galleryClass, payload, id) {
+    let banners = payload.product_banners.map(original => {
+      let banner = {
+        guid: original.id
```
why are you shadowing the `id` as a `guid`?

walsh9 replied:
i thought ember models had their own `id` property, usually the lookup key... don't want to conflict with that, but actually not sure how ember-data-model-fragments handles this

walsh9 replied:
Looks like ember-data-model-fragments doesn't use anything like an id field so yeah i can just call this `id`

walsh9 replied:
okay actually:
`You may not set 'id' as an attribute on your model. Please remove any lines that look like: "id: DS.attr('<type>')" from model:product-banner`

noslouch commented on app/templates/index.hbs:
```diff
@@ -3,32 +3,37 @@
+    {{#if (and
+      (not topProductBannerDismissed)
+      (not (cookie-exists productBanner.cookieId))
+      (eq productBanner.location 'TOP')
```
in fact, since you're creating ember models for all of the product banners, why not let each banner hold its own state? something like `{{#each model.systemMessage.visibleBanners ...}}` might be nicer than including all these conditionals

noslouch commented on app/controllers/index.js:
```diff
@@ -66,10 +58,12 @@ export default Controller.extend({
+    hideBannerForHours(cookie, hours, location) {
+      let expires = moment().add(hours, 'hours').toDate();
+      this.cookies.write(cookie, 1, {expires, path: '/'});
+      if (location === "TOP") {
```
also think about refactoring this action to accept the product banner model instead of these various scalar values. as new banners get added, I could see this action getting more complicated. I think instead it should do two things: 1) set the cookie 2) call `dismiss` on a product banner model. maybe it even makes sense to have the product banner model set the cookie, so the controller is less involved?

noslouch commented on tests/integration/helpers/cookie-exists-test.js:
```diff
@@ -0,0 +1,17 @@
+    await render(hbs`{{cookie-exists inputValue}}`);
+
+    assert.equal(this.element.textContent.trim(), 'false');
+  });
```
can you include a test for this helper?

walsh9 commented:
- Using ember-data's build in `EmbeddedRecordsMixin` instead of `ember-data-model-fragments`, this handles sub-model's ids better
- Handling banner visibility on the model makes the template part a lot simpler
- Got mirage factories working nicely, writing tests should be easier now

noslouch commented on app/models/product-banner.js:
```diff
@@ -0,0 +1,24 @@
+  isCookieSet: computed('cookieId', function() {
+    return this.cookies.exists(this.cookieId);
+  }),
+  isDismissed: attr('boolean', {defaultValue() {return false;}}),
```
if it's a scalar value, you don't need a method. you can specify with a pojo: `attr('boolean', {defaultValue: false})`

noslouch replied:
in fact... won't this default to a falsey value anyway?

noslouch commented on mirage/factories/product-banner.js:
```diff
@@ -0,0 +1,32 @@
+export default Factory.extend({
+  id() {
+    return faker.random.uuid();
+  },
```
prefer arrow functions for these factory methods if they're one liners: `id: () => faker.random.uuid()`

noslouch commented on mirage/factories/product-banner.js:
```diff
@@ -0,0 +1,32 @@
+  id() {
+    return faker.random.uuid();
+  },
+  "type": "product-banner",
```
strip the quotes if they're not needed

noslouch commented on app/controllers/index.js:
```diff
@@ -66,10 +57,10 @@ export default Controller.extend({
+    hideBanner(productBanner) {
+      let expires = moment().add(productBanner.frequency, 'hours').toDate();
+      this.cookies.write(productBanner.cookieId, 1, {expires, path: '/'});
+      productBanner.set('isDismissed', true);
```
this could be taken as code geography, but I think given that the product banner is already injecting the cookie service, I'd like to see this method on the product banner model. I'm thinking about using this banner component/model combo in other places, and it'd be nice if we didn't need to also add some code to a backing controller or component file in order for it to work. 

If the method goes on the model, then all the logic for showing/hiding is encapsulated into the model itself, and you can use attrs off the model to decide on the template whether or not to show the message. what do you think?

noslouch replied:
and then can you include a test for that method on the model's unit test as well?

noslouch commented on tests/unit/serializers/system-messages-test.js:
```diff
@@ -0,0 +1,23 @@
+
+    assert.ok(serializedRecord);
+  });
+});
```
can you add a test for the system message serializer?

walsh9 commented:
Agree that the cookie dismiss behavior shouldn't be on the route controller, but moving it to the model felt weird when I tried to use it*.  After considering a few alternatives i've moved this UI behavior to a `DismissableBlock` wrapper component which generically handles dismissing sections and setting and checking cookies.

Also:
- Cleaned up the product banner mirage factory
- Added tests to the system messages serializer

(*One example of why it felt weird: models don't have actions, and using a plain function doesn't bind the model instance itself to `this`, so you end up needing to do something like `{{action productBanner.hide productBanner}}` when it feels like you should be able to just do  `{{action productBanner.hide}}` which won't work.)

walsh9 commented:
the `DismissableBanner` component is generic enough that we'll probably want to extract it into one of our addons, but for now let's just consider using it in this client as a beta test
